src/libstd/sys/unix/fs.rs

```diff
@@ -251,21 +251,35 @@ impl Iterator for ReadDir {
 
     #[cfg(not(any(target_os = "solaris", target_os = "fuchsia")))]
     fn next(&mut self) -> Option<io::Result<DirEntry>> {
-        unsafe {
-            let mut ret = DirEntry {
-                entry: mem::zeroed(),
-                root: self.root.clone()
-            };
-            let mut entry_ptr = ptr::null_mut();
-            loop {
-                if readdir64_r(self.dirp.0, &mut ret.entry, &mut entry_ptr) != 0 {
-                    return Some(Err(Error::last_os_error()))
-                }
-                if entry_ptr.is_null() {
-                    return None
-                }
-                if ret.name_bytes() != b"." && ret.name_bytes() != b".." {
-                    return Some(Ok(ret))
+        if self.dirp.is_none() {
```

Comment:
Don't we make this check again just below on 259?

Reply:
Oh, absolutely - I will remove this `if` clause.